### DISCUSSION TITLE
Cannot start a runtime from within a runtime.

### DISCUSSION THREAD

**Discussion by sagarnayak** [rwf2/Rocket#1754] [Questions]
Api call -

	use rocket::serde::json::Json;
	use rocket::State;

	use crate::core::database::db_pool::DbPool;
	use crate::model::status_message::StatusMessage;
	use crate::model::user::User;

	#[get("/getUsersFromDatabase")]
	pub fn get_all_user(db_pool: &State<DbPool>) -> Result<Json<Vec<User>>, Json<StatusMessage>> {
		let mut db = db_pool.database.get().unwrap();

		let statement = db.prepare(
			"SELECT * FROM users"
		);

		match statement {
			Err(_) => {
				return Err(
					Json(
						StatusMessage {
							code: 400,
							message: "test error".to_string(),
						}
					)
				);
			}
			_ => {}
		}

		let result_for_users = db.query(&statement.unwrap(), &[]);

		match result_for_users {
			Err(_) => {
				return Err(
					Json(
						StatusMessage {
							code: 400,
							message: "test error".to_string(),
						}
					)
				);
			}
			_ => {}
		}

		Ok(
			Json(
				vec![
					User {
						name: "test user".to_string()
					}
				]
			)
		)
	}

db pool creation code -

	use postgres::NoTls;
	use r2d2_postgres::PostgresConnectionManager;
	use r2d2_postgres::r2d2::Pool;

	use crate::config_controller::ConfigData;
	use crate::core::database::db_pool::DbPool;

	fn get_pool() -> Pool<PostgresConnectionManager<NoTls>> {
		let config = ConfigData::new().expect("Failed to get config data");
		let database = config.database;

		let host = database.host;
		let port = database.port;
		let user = database.user;
		let password = database.password;
		let pool_size = database.pool_size;

		let manager = PostgresConnectionManager::new(
			format!("host={} port={} user={} password={}", host, port, user, password).parse().unwrap(),
			NoTls,
		);

		println!("created database manager");

		Pool::builder().max_size(pool_size).build(manager).unwrap()
	}

	pub fn get_db_pools() -> DbPool {
		DbPool {
			database: get_pool()
		}
	}

I am getting this error when i call the above end point -

	   >> Matched: (get_all_user) GET /getUsersFromDatabase
	thread 'rocket-worker-thread' panicked at 'Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.', C:\Users\SAGAR\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.8.1\src\runtime\enter.rs:39:9
	note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
	thread 'rocket-worker-thread' panicked at 'Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.', C:\Users\SAGAR\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.8.1\src\runtime\enter.rs:39:9

Please do suggest what am I missing.

**Comment by jebrosen** [comment]
Since 0.5, Rocket uses Rust's async/await programming with [Futures](https://rocket.rs/v0.5-rc/guide/overview/#futures-and-async).

For postgres, you should prefer either the [existing support for `postgres` via `rocket_sync_db_pools`](https://rocket.rs/v0.5-rc/guide/state/#databases), which makes the non-async `postgres` API usable from async code, or use an async-first database connection pool such as [`deadpool-postgres`](https://crates.io/crates/deadpool_postgres) instead of `r2d2`. #1117 is the tracking issue for making an async equivalent for `rocket_sync_db_pools`.

**Reply by sagarnayak** [reply]
Hi, any recent example of using rocket with Postgres without any ORM. I am new to rust and unable to find any recent example.

**Reply by jebrosen** [reply]
We don't currently provide such an example in this repository, but it's fairly simple to do with `deadpool-postgres` or other crates. <https://www.reddit.com/r/rust/comments/e9n6mx/creating_and_using_a_postgres_database_pool_is/> demonstrates the overall approach - initialize a connection `Pool` and put into managed state, and retrieve connections in routes or request guards as needed.

**Comment by snailwei** [comment]
Found a solution here
https://stackoverflow.com/a/62536772/2258077
